man/boot-block.1
.TH "boot-block" 1 "February 15, 2012" "version 0.99.19" "COBALT COMMANDS"
.SH NAME
boot-block \- Prompt Cobalt to execute commands on a BlueGene compute block on the user's behalf
.SH SYNOPSIS
.B boot-block
[ --block
.I block-name
] [ --jobid
.I jobid
] [--free] [--reboot]
.SH DESCRIPTION
.PP
.B boot-block
is a command that prompts Cobalt's system component to boot, free, or reboot a
a block on behalf of a user.  This can boot, free or reboot a block.  The command will complete after
the requested operation has completed.  To execute commands on a block, the user issuing the command
must be the user running the job, and must have that block, or a block with a superset of the resources
allocated to them through Cobalt.  Once the Cobalt resource allocation ends, all blocks booted by this
command on the allocated block will be freed.
.SH OPTIONS
.TP
.B \-d \--debug
Turn on communication debugging
.TP
.B \-h \-\-help
Displays the usage and a brief descriptions of the options
.TP
.B \-\-version
Displays client revision and Cobalt version
.TP
.B \-\-block block-name
Boot the specified block. If the block is already booted, this will not cause
the block to free.  If this is not specified, the COBALT_PARTNAME
enviornment variable is checked and used as the block to boot.
.TP
.B \-\-jobid jobid
Boot a block for the specified jobid.  The executing user must be the owner of the
resource allocation in Cobalt, and the jobid must be running.  The environment variable
COBALT_JOBID is used if this option is not provided.
.TP
.B \-\-free
Free the specified block, and allow it's resources to be rebooted, or be rebooted as a part
of another block.  All runjobs on the target resource must be terminated for a block free
to complete. This must not be used in conjunction with the
.BR \-\-reboot
flag.
.TP
.B \-\-reboot
Reboot the specified block.  This executes a free followed by a boot.  The block will be
ready to run jobs on completion.  Like the
.BR \-\-free
option the free stage will only
complete if all running jobs on the target block have been terminated.  This may not be used
with the
.BR \-\-free
option.

.SH "SEE ALSO"
.BR qsub (1),
.BR get-bootable-blocks (1)


CHANGES
= Changes from previous Cobalt Versions =


== Changes to 0.99.27 ==
 * Alternate kernel support for BG/Q compute nodes and IO nodes is available.
  Numerous configuration settings have been added to support this.
  (Cobalt ticket 621)
 * If $jobid is included in a job name specification or an environment variable,
  at submission time, Cobalt will substitute the generated jobid for that job.
  See man qsub for more details. (Cobalt ticket 617)
 * The Cobalt jobid is now explicitly added in the cobaltlog file. (Cobalt ticket 390)
 * The name of a job can now be specified seperately from naming the output
  files. (Cobalt ticket 625)
 * Multiple locations can be passed to reservations through the -p option as a
  ':'-delimited list. (Cobalt ticket 405)
 * cqadm can now be used to release user hold, as well as put jobs into a user hold
  as well as an admin hold (Cobalt ticket 382)
 * The users field can be removed from a reservation using setres -m -u '*'. (Cobalt
   ticket 381)
 * Releaseres should now report the correct partiton count on reservation release
  (Cobalt ticket 470)
 * Reservations now show time remaining in showres once the reservation is active.
  (Cobalt ticket 585)
 * Schedctl should handle flipping score and jobid arguments more gracefully and 
  no longer silently discards invalid options. (Cobalt ticket 502)
 * ':' and '=' may now be escaped as '\:' and '\=' in the --envs argument (Cobalt
  ticket 589)
 * Reservations may now take 'now' for the start time of a reservation (Cobalt
  ticket 546)
 * "GEOMETRY", "ION_KERNEL" and "ION_KERNELOPTIONS have been added to the JOB_DATA
  table of the Cobalt database 


== Changes to 0.99.26 ==
 * Corrected a bug that was causing messages in the clients that used to go to
  stdout to go to stderr instead

== Changes to 0.99.25 ==
 * Fixes for clients that were causing erronious default values for proccount
  have been fixed.
 * The setgid wrapper now calls a site-selectable python interperter with the
  '-E' flag by default.
 * A bug that could cause TimeRemaining to show negative values has been fixed
 * Queues may be added or removed from blocks without overwriting the entire
  list with new options in partadm using --rmq or --appq with the --queue option.


== Changes to 0.99.24 ==
 * Reverting client changes.  Found some issues when deploying to test systems
 * --defer flag has been reverted with these as a casualty.

== Changes to 0.99.23 ==
 * Client code has undergone a refactor.  Numerous bugs have been addressed
  with these that were found while improving test coverage of this code
 * IO Block booting and automatic control has been added.  Please consult the
  partadm manpage for details
 * This release requires pybgsched-0.1.3
 * A user may voluntarily set their job's score to 0 and let it re-accrue using
  qalter --defer (Ticket #590)

== Changes to 0.99.22 ==
 * Fix for a deadlock that could occur due to a race condition in the 
  boot-reaping code.
 * Fix for a pid collision that could occur after a forced delete of a job in
  the system-script-forker component leading to a job getting stuck in the
  Job_Epilogue state.
 * python 2.7 and it's change to the XMLRPC libraries are now supported.

== Changes to 0.99.21 ==
 * Minor fix release to correct a failure mode where we can lose a process group
  if a job's boot is delayed.

== Changes to 0.99.20 ==
 * Small changeset to fix a race condition that was hanging jobs that came in
   on 0.99.19
 * Better error logging for situations where the boots and the process group
  information haven't lined up yet.  This usually happens during high-load and
  can result in block boot completion never being detected.

== Changes to 0.99.19 ==
 * Manual boot control has been added
 * major refactor of compute block rebooting

== Changes to 0.99.18 ==
 * Fix for other systems that were seeing erronious geometry data
 * SoftwareFailure now accurately reported

== Changes to 0.99.17 ==

 * Support for --geometry has been added.  This will constrain a job to run 
blocks having a specific geometry in nodes, partadm, partlist and qstat have
been updated to show this.
 * Support for passthrough blocking reservations has been added.
 * Fixed a bug where a job run outside of Cobalt would not properly block resources.
 * Removed autoreboot in the event of a block entering a control system 
  SoftwareFalure state on BG/Q.  Normal cleanup should handle this now.

== Changes to 0.99.15 ==
 * Fix for an issue that was causing partadm to malfunction when reservations
were set on the system.

== Changes to 0.99.12 ==
 * As a note, this is a release of fixes and backported features from BG/Q
 * Fixed issues that could cause a bad hardware state to be overridden as 
"blocked" allowing the block to be drained
 * Fix for a situation where a partition that is inelligible to run due to a
pending reservation is selected for draining.

== Changes to 0.99.11 ==
 * Fix for an additional circumstance that can cause a small reservation to
collapse a backfill window to zero.

== Changes to 0.99.10 ==
 * Fixed an issue that was introduced in 0.99.6 where children were not being
properly added to a block.  This caused strange behavior in reservations,
including failure to run jobs smaller than the reservation or a small job
running on the largest block in a reservation
 * Refactored the update_block_state code to operate more similarly to the way
BG/P is currently doing it.  This should correct an issue where a block can
have an error state overridden.
 * Cleaned up a circumstance where wires between passthrough midplanes can be
added to a block's wire list.  This can cause erronious wiring conflicts to be
detected.
 * Fixed an issue where duplicate wires were being added

== Changes to 0.99.9 ==

 * Fixed an issue where a job could choose a block for draining that is blocked
by a pending reservation.  This would cause the backfill time to collapse to zero.

src/clients/boot-block.py
#!/usr/bin/env python
"""
Prompt cobalt to boot a block on behalf of a user.

Usage: %prog [options]
version: "%prog " + __revision__ + , Cobalt  + __version__

OPTIONS DEFINITIONS:

'-d','--debug',dest='debug',help='turn on communication debugging',callback=cb_debug
'--block', dest='block', type='string', help='Name of block to boot.'
'--reboot', action='store_true', dest='reboot', help='If the block is already booted, free the block and reboot.'
'--free', action='store_true', dest='free', help='Free the block, if booted.  May not be combined with reboot'
'--jobid', dest='jobid', type='int', help='Specify a cobalt jobid for this boot.',callback=cb_gtzero

"""
import logging
import sys
import os
from Cobalt import client_utils
from Cobalt.client_utils import cb_debug, cb_gtzero

from Cobalt.arg_parser import ArgParse

__revision__= 'TBD'
__version__ = '$Version$'

AUTH_FAIL       = 2
BAD_OPTION_FAIL = 3
GENERAL_FAIL    = 1
SUCCESS         = 0

SYSMGR = client_utils.SYSMGR

def main():
    """
    get-bootable-blocks main
    """
    # setup logging for client. The clients should call this before doing anything else.
    client_utils.setup_logging(logging.INFO)

    # list of callback with its arguments
    callbacks = [
        # <cb function>     <cb args>
        [ cb_debug        , () ],
        [ cb_gtzero       , (True,) ] ] # return int

    # Get the version information
    opt_def =  __doc__.replace('__revision__',__revision__)
    opt_def =  opt_def.replace('__version__',__version__)

    parser = ArgParse(opt_def,callbacks)

    user = client_utils.getuid()

    parser.parse_it() # parse the command line
    opts   = parser.options

    if not parser.no_args():
        client_utils.logger.info('No arguments needed')

    if opts.free and opts.reboot:
        client_utils.logger.error("ERROR: --free may not be specified with --reboot.")
        sys.exit(BAD_OPTION_FAIL)

    block = opts.block
    if block == None:
        try:
            block = os.environ['COBALT_PARTNAME']
        except KeyError:
            pass
        try:
            block = os.environ['COBALT_BLOCKNAME']
        except KeyError:
            pass
        if block == None:
            client_utils.logger.error("ERROR: block not specified as option or in environment.")
            sys.exit(BAD_OPTION_FAIL)

    jobid = opts.jobid
    if jobid == None:
        try:
            jobid = int(os.environ['COBALT_JOBID'])
        except KeyError:
            client_utils.logger.error("ERROR: Cobalt jobid not specified as option or in environment.")
            sys.exit(BAD_OPTION_FAIL)

    if opts.reboot or opts.free:
        #Start the free on the block
        #poke cobalt to kill all jobs on the resource as well.
        success = client_utils.component_call(SYSMGR, False, 'initiate_proxy_free', (block, user, jobid))
        client_utils.logger.info("Block free on %s initiated." % (block,))
        if not success:
            client_utils.logger.error("Free request for block %s failed authorization." % (block, ))
            sys.exit(AUTH_FAIL)
        while (True):
            #wait for free.  If the user still has jobs running, this won't complete.
            #the proxy free should take care of this, though.
            if client_utils.component_call(SYSMGR, False, 'get_block_bgsched_status', (block,)) == 'Free':
                client_utils.logger.info("Block %s successfully freed." % (block,))
                break

    if not opts.free:
        success = client_utils.component_call(SYSMGR, False, 'initiate_proxy_boot', (block, user, jobid))
        if not success:
            client_utils.logger.error("Boot request for block %s failed authorization." % (block, ))
            sys.exit(AUTH_FAIL)
        #give the system component a moment to initiate the boot
        client_utils.sleep(3)
        #wait for block to boot
        failed = False
        found = False
        while True:
            boot_id, status, status_strings = client_utils.component_call(SYSMGR, False, 'get_boot_statuses_and_strings', (block,))
            if not found:
                if boot_id != None:
                    found = True
            else:
                if status_strings != [] and status_strings != None:
                    print "\n".join(status_strings)
                if status in ['complete', 'failed']:
                    client_utils.component_call(SYSMGR, False, 'reap_boot', (block,))
                    if status == 'failed':
                        failed = True
                    break
            client_utils.sleep(1)
        if failed:
            client_utils.logger.error("Boot for locaiton %s failed."% (block,))
        else:
            client_utils.logger.info("Boot for locaiton %s complete."% (block,))

if __name__ == '__main__':
    try:
        main()
    except SystemExit:
        raise
    except Exception, e:
        client_utils.logger.fatal("*** FATAL EXCEPTION: %s ***", e)
        sys.exit(1)
